staging/src/k8s.io/apiserver/pkg/audit/request.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package audit
+
+import (
+	"net/http"
+	"strings"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authenticationv1 "k8s.io/client-go/pkg/apis/authentication/v1"
+)
+
+func UpdateEventFromRequest(ctx request.Context, ev *auditinternal.Event, req *http.Request, attribs authorizer.Attributes) error {
+	ev.Verb = attribs.GetVerb()
+	ev.RequestURI = req.URL.RequestURI()
+	ev.SourceIP = utilnet.GetClientIP(req).String()
+
+	if user := attribs.GetUser(); user != nil {
+		ev.User.Username = user.GetName()
+		ev.User.Extra = map[string]auditinternal.ExtraValue{}
+		for k, v := range user.GetExtra() {
+			ev.User.Extra[k] = auditinternal.ExtraValue(v)
+		}
+		ev.User.Groups = user.GetGroups()
+		ev.User.UID = user.GetUID()
+	}
+
+	if asuser := req.Header.Get(authenticationv1.ImpersonateUserHeader); len(asuser) > 0 {
+		ev.Impersonate = &auditinternal.UserInfo{
+			Username: asuser,
+		}
+		if requestedGroups := req.Header[authenticationv1.ImpersonateGroupHeader]; len(requestedGroups) > 0 {
+			ev.Impersonate.Groups = requestedGroups
+		}
+
+		ev.Impersonate.Extra = map[string]auditinternal.ExtraValue{}
+		for k, v := range req.Header {
+			if !strings.HasPrefix(k, authenticationv1.ImpersonateUserExtraHeaderPrefix) {
+				continue
+			}
+			k = k[len(authenticationv1.ImpersonateUserExtraHeaderPrefix):]
+			ev.Impersonate.Extra[k] = auditinternal.ExtraValue(v)
+		}
+	}
+
+	if attribs.IsResourceRequest() {
+		ev.ObjectRef = &auditinternal.ObjectReference{
```

Comment:
In the URI you have the resource, not the kind. It's what you gets from `attribs`.